### PR TITLE
Wait for submissions to complete on `Queue` drop

### PR DESCRIPTION
**Pull request by teoxoy**
This PR addresses leaks caused by circular references due to the `Device` still holding strong references to resources when removed from the registry and while submissions are still active. `Global::device_drop` was wrongly assuming `device_poll` with `Maintain::Wait` was called but this is not a documented invariant and only `wgpu` was upholding this.

Instead of calling `device_poll` in `device_drop` (which would solve the issue) this PR takes a different approach since we want to remove the registries in the future (https://github.com/gfx-rs/wgpu/issues/5121):

- it moves the `LifetimeTracker` and `PendingWrites` into the `Queue` so that we never have circular references
- it adds the wait for active submissions in `Queue`'s `Drop` impl

One thing to note is that ideally we shouldn't be waiting in the `Queue`'s `Drop` implementation but that's what `wgpu` was previously doing (by calling `device_poll`) and the alternative is more involved: We could put the burden of keeping the `Queue` alive on `wgpu-core` users if there are any active submissions. With the changes in this PR we will panic if we hit a timeout or ran into any errors; this would solve that too. I want to talk about this approach at the next maintainers call before making any changes though.

### CODE REVIEW

**Comment by teoxoy** [comment]
CI was failing because when targeting WebGL we were timing out. eade5cfb4027095accec7ca31af1b3a00872aaa3 fixes CI and behaves the same as current trunk where the timeout is ignored (related: https://github.com/gfx-rs/wgpu/issues/3601 & https://github.com/gfx-rs/wgpu/issues/4589) but is not correct.

This is another argument for requiring users of `wgpu-core` to keep the `Queue` alive and keep polling it until submissions complete.

**Comment by teoxoy** [comment]
Conclusions after discussing the idea of requiring users of `wgpu-core` to keep the `Queue` alive and keep polling it until submissions complete with the other maintainers:
- Most users don’t call poll so we shouldn’t make it a requirement (maintain gets called implicitly on submit).
- @cwfitzgerald said it should be fine to drop everything early on WebGL since WebGL implementations will still keep the objects alive (I will double check with @kdashg).
- We should still implement the idea for Firefox since we shouldn't be blocking the GPU process.

**Comment by teoxoy** [comment]
Some follow-up work is still needed though, instead of panicking, we need to add a retry mechanism if we time out or if we run into OOMs. Device loss needs to be propagated to the device (making it invalid).

**Comment by teoxoy** [comment]
I revised the comment explaining why it's actually ok that we time out on WebGL.

**Comment by teoxoy** [comment]
I added the retry mechanism as well, this should be ready for review.

**Comment by ErichDonGubler** [comment]
Looks like this has compilation issues after being rebased onto #6291.

**Comment by teoxoy** [comment]
The last commit (see its description for details) fixes the issue that CI was blocked on in https://github.com/gfx-rs/wgpu/actions/runs/11799523804/job/32869472729

I have a follow-up that cleans up command buffers/encoders since they are currently more complicated than they can be; will put up the PR once this one lands.

File: wgpu-core/src/device/resource.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/resource.rs]: thought: I suspect that moving the `device_lost_closure` call to `Drop` will resolve some issues I've seen in CTS indicating that it wasn't expecting to see the device lost callback being invoked.

File: wgpu-core/src/resource.rs
Comment (ErichDonGubler) [inline wgpu-core/src/resource.rs]: thought: This and the below changes are interesting, because previously, `device.pending_writes` had a lock region that overlapped the `life_lock` region. Now, they're not overlapping at all.

Comment (ErichDonGubler) [reply wgpu-core/src/resource.rs]: question: Actually, I'm surprised this doesn't result in a deletion of an entry in the `QUEUE_PENDING_WRITES` entry in `wgpu-core/src/lock/rank.rs`. I'm guessing that this wasn't done because there are other cases where the `life_lock` is acquired within the region of a `pending_writes` lock?

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/queue.rs:46]: question: RE: the `move PendingWrites into the Queue` commit: The commit message claims that it avoids (removes?) strong ref. cycles between a `Device` and resources containing backlinks to it. I don't see how moving `pending_writes` from a `Device` to a `Queue` (which seems to be held by a `Device`, and also hold a strong ref. to a `Device) does anything other than increase the number of steps in a cycle. Am I missing something?

Comment (ErichDonGubler) [reply wgpu-core/src/device/queue.rs:46]: Ah, I think I understand the answer: a `Queue` is held behind a `Weak` ref., rather than an `Arc`, in a `Device`. So, this should break a cycle, because the ref. link to `Queue` is comparatively weaker.

File: wgpu-core/src/device/resource.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/resource.rs:79]: note: CC @jimblandy for the ` Remove outdated locking comments ` commit specifically.

File: wgpu-core/src/device/global.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/global.rs:2101]: suggestion(non-blocking): I think the ` move device_lost_closure on the Device ` commit title is less clear than, say, ``move `device_lost_closure` from `Device::life_tracker`  to `Device``

File: wgpu-core/src/device/global.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/global.rs]: nitpick: This seems like a re-invention of `Option::replace`:


```suggestion
        let old_device_lost_closure = device
            .device_lost_closure
            .lock()
            .replace(device_lost_closure);
```

File: wgpu-core/src/device/resource.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/resource.rs]: question: This change eliminates coverage of a `life_tracker` lock region over some cleanup operations after this hunk. I presume this is safe (from a place of ignorance), because these should be exclusively held resources. Is that right? I'm nervous about the `fence` in particular, because we acquire a `write` lock on it before `ManuallyDrop::take`ing it.

File: wgpu-core/src/resource.rs
Comment (ErichDonGubler) [inline wgpu-core/src/resource.rs]: question: Is there a reason we can't immediately implement `TODO: map immediately` in the ` move LifetimeTracker into the Queue ` commit?

I'll likely see the answer to this myself shortly, since I'm not all the way through commits, and I see this `TODO` doesn't survive in the overall diff.

File: wgpu-core/src/device/mod.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/mod.rs]: nitpick: It seems like this only needs to be promoted to `pub(crate)`? 🤔

File: wgpu-core/src/resource.rs
Comment (ErichDonGubler) [reply wgpu-core/src/resource.rs]: Yep, answered in the next commit. 🫡

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/queue.rs:149]: todo: I want to double-check atomic orderings again, just to make sure I feel confident with this. I don't trust myself to eyeball this. 😅

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/queue.rs]: note: panicking in `Drop` can result in an abortion of the program if `Drop` was being called in the middle of a panic (see [`std::Drop`'s `Panics` section](https://doc.rust-lang.org/std/ops/trait.Drop.html#panics)).

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/queue.rs]: nitpick: Missing a `SAFETY` comment here.

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/queue.rs]: nitpick: Manual `drop`s of locks are sometimes necessary, but this doesn't seem like a case where it _is_ necessary.	

suggestion: I think it's strictly easier to perform assignment of these variables inside a block, so we can't forget to unlock locks.

```suggestion
        let wait_res;
        {
            let fence = self.device.fence.read();
            wait_res = unsafe {
                self.device.raw().wait(
                    fence.as_ref(),
                    last_successful_submission_index,
                    crate::device::CLEANUP_WAIT_MS,
                )
            };
        }

        match wait_res {
            Ok(true) => {}
            // Note: If we don't panic here we are in UB land (destroying resources while they are still in use by the GPU).
            Ok(false) => {
                panic!("We timed out while waiting on the last successful submission to complete!");
            }
            Err(e) => {
                panic!(
                    "We ran into an error while waiting on the last successful submission to complete! - {e}"
                );
            }
        }

        let (submission_closures, mapping_closures, queue_empty);
        {
            let snatch_guard = self.device.snatchable_lock.read();
            (submission_closures, mapping_closures, queue_empty) =
                self.maintain(last_successful_submission_index, &snatch_guard);
        }
```

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/queue.rs]: todo: I don't think we should leave a `TODO` here on the final merge; we can just delete these lines after we file an issue.

File: wgpu-hal/src/gles/device.rs
Comment (ErichDonGubler) [inline wgpu-hal/src/gles/device.rs:1562]: question: How is the above commentary relevant to the `cfg!(…)` below? These are user agents, but we're not branching on anything like a user agent report.

File: wgpu-hal/src/gles/mod.rs
Comment (ErichDonGubler) [inline wgpu-hal/src/gles/mod.rs:301]: question: Why was the body of `gles::Device` not migrated in its entirety here?

answer: Oh, I see that it got split between `impl Drop for Device` and `impl Drop for Queue`.

question: This means that these operations were re-ordered a bit. Is this okay?

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/queue.rs:220]: question: If we panic here, but don't abort, then we may have poisoned the `fence` lock here. Should we `drop` it before panicking?

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [inline wgpu-core/src/device/queue.rs:1214]: thought: I guess that this is technically more correct than before. The diagnostics seem like they would be wrong if a user attempts to use these command buffers again, though (i.e., "You can't use this, this is destroyed." as opposed to "this is invalid").

Not concerned either way, though.

**Review by ErichDonGubler — COMMENTED** [review]
Inclined to approve if we can get my questions answered.

File: wgpu-core/src/resource.rs
Comment (teoxoy) [reply wgpu-core/src/resource.rs]: Maybe, but I didn't check.

Comment (teoxoy) [reply wgpu-core/src/resource.rs]: One of the next commits that moves the `LifetimeTracker` into the `Queue` changes this back to the previous shape.

File: wgpu-core/src/device/resource.rs
Comment (teoxoy) [reply wgpu-core/src/device/resource.rs]: As far as I understand in this case the lifetime of the guard didn't have to be extended, it's in a temporary scope.

File: wgpu-core/src/device/queue.rs
Comment (teoxoy) [reply wgpu-core/src/device/queue.rs]: I guess I should be wrapping them in `if !thread::panicking() {`?

File: wgpu-core/src/device/queue.rs
Comment (teoxoy) [reply wgpu-core/src/device/queue.rs]: I don't know what the safety requirements of calling this function are... I'd have to check and document all the backends for this. All wgpu-hal functions are marked as unsafe without a safety contract. We should document the safety contract at some point but I don't think this is the time for that.

File: wgpu-hal/src/gles/device.rs
Comment (teoxoy) [reply wgpu-hal/src/gles/device.rs:1562]: It's info as to why the timeout needs to be 0 on WebGL. I was confused at first why that was given that there was `MAX_CLIENT_WAIT_TIMEOUT_WEBGL` that apps could query for but it turns out to be a static 0 almost everywhere.

File: wgpu-hal/src/gles/mod.rs
Comment (teoxoy) [reply wgpu-hal/src/gles/mod.rs:301]: I checked and it's fine.

File: wgpu-core/src/device/queue.rs
Comment (teoxoy) [reply wgpu-core/src/device/queue.rs:220]: This is a `parking_lot::RwLock` which says "No poisoning, the lock is released normally on panic.".

File: wgpu-core/src/device/queue.rs
Comment (teoxoy) [reply wgpu-core/src/device/queue.rs:1214]: From the user's POV command buffers can't be destroyed. They always become invalid once submitted. Here we were just forgetting to destroy the underlying command buffer/encoder.

File: wgpu-core/src/device/queue.rs
Comment (teoxoy) [reply wgpu-core/src/device/queue.rs]: I prefer to have the explicit `drop` calls rather than adding new scopes to trigger `drop` implicitly.

File: wgpu-core/src/device/queue.rs
Comment (teoxoy) [reply wgpu-core/src/device/queue.rs]: Actually I think aborting in this case is appropriate. If we don't, that's UB and we shouldn't continue unwinding.

File: wgpu-core/src/device/queue.rs
Comment (teoxoy) [reply wgpu-core/src/device/queue.rs]: Filed: https://github.com/gfx-rs/wgpu/issues/6538

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [reply wgpu-core/src/device/queue.rs]: [Definitely not blocking](https://web.archive.org/web/20241002184501/https://conventionalcomments.org/#labels), by any means. Running into an abort here means (1) a user was panicking for some reason and (2) an error was returned in a function that has no means of dealing with an error.

If somebody runs into errors, do we have a clear workaround identified? If so, I'm a _lot_ less concerned about this, though I will not block even if there is not (since I know there's yet more follow-up work to do that should resolve it anyway).

File: wgpu-core/src/device/queue.rs
Comment (ErichDonGubler) [reply wgpu-core/src/device/queue.rs:149]: Seems right to me, resolving!

**Comment by ErichDonGubler** [comment]
LGTM, with the discussions we've had. If you're still interested in @jimblandy's feedback, I'd suggest giving 'im...maybe 24 hours, if you're feeling patient? Otherwise, I think it's reasonable to do review after merging, if really necessary.

**Review by ErichDonGubler — APPROVED** [review]

File: wgpu-core/src/device/queue.rs
Comment (teoxoy) [reply wgpu-core/src/device/queue.rs]: I implemented a retry mechanism for the cases that we can recover from. I can't think of other ways to deal with these errors.

**Comment by teoxoy** [comment]
Let's land it, I already had to rebase it and fix conflicts a few times.

**Pull request merged**
